parla/cuda.py

```diff
@@ -163,27 +162,68 @@ def get_array_module(self, a):
 
 # Integration with parla.environments
 
+class _GPUStacksLocal(threading.local):
+    _stream_stack: List[cupy.cuda.Stream]
+    _device_stack: List[cupy.cuda.Device]
+
+    def __init__(self):
+        super(_GPUStacksLocal, self).__init__()
+        self._stream_stack = []
+        self._device_stack = []
+
+    def push_stream(self, stream):
+        self._stream_stack.append(stream)
+
+    def pop_stream(self) -> cupy.cuda.Stream:
+        return self._stream_stack.pop()
+
+    def push_device(self, dev):
+        self._device_stack.append(dev)
+
+    def pop_device(self) -> cupy.cuda.Device:
+        return self._device_stack.pop()
+
+    @property
+    def stream(self):
+        if self._stream_stack:
+            return self._stream_stack[-1]
```

Comment:
Just to confirm that I'm understanding right, the only reason this won't cause unintentional dependencies between GPU kernel calls is that whichever task most recently used a given stream already did stream.synchronize before pushing the stream back onto the stack?

Reply:
This property is actually never used. I meant to remove it.

Also streams are never reused. They are created, entered, pushed, popped, exited, then discarded. They are never shared.

Reply:
Excellent. Supposedly stream creation overhead is extremely small, so I'd much rather just run with a new one.

Reply:
It turns out that I need this function to implement the ordering you mention, but the semantics I stated are still true.